compiler/src/iree/compiler/GlobalOptimization/FoldGlobalUnitDims.cpp

```diff
@@ -0,0 +1,156 @@
+  if (!expandShapeReInds) {
+    return failure();
+  }
+
+  for (auto load : loadOps) {
+    rewriter.setInsertionPoint(load);
+    auto newOp =
+        rewriter.create(load->getLoc(), load->getName().getIdentifier(),
+                        load->getOperands(), {newGlobalType}, load->getAttrs());
+    auto newLoad = dyn_cast<IREE::Util::GlobalLoadOpInterface>(newOp);
+    newLoad.setGlobalAttr(FlatSymbolRefAttr::get(newGlobal.getGlobalName()));
+    newLoad.setGlobalImmutable(load.isGlobalImmutable());
+    rewriter.replaceOpWithNewOp<tensor::ExpandShapeOp>(
+        load, globalType, newLoad->getResult(0), expandShapeReInds.value());
+  }
+  for (auto store : storeOps) {
+    rewriter.setInsertionPoint(store);
+    Value collapse = rewriter.create<tensor::CollapseShapeOp>(
+        store.getLoc(), newGlobalType, store->getOperand(0),
+        expandShapeReInds.value());
+    auto newOp = rewriter.create(
+        store->getLoc(), store->getName().getIdentifier(), store->getOperands(),
+        store->getResultTypes(), store->getAttrs());
+    auto newStore = dyn_cast<IREE::Util::GlobalStoreOpInterface>(newOp);
+    newStore.setGlobalAttr(FlatSymbolRefAttr::get(newGlobal.getGlobalName()));
+    newStore.setStoredGlobalValue(collapse);
+    rewriter.eraseOp(store);
+  }
+  rewriter.eraseOp(global);
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Pass definition
+//===----------------------------------------------------------------------===//
+
+class FoldGlobalUnitDimsPass
+    : public FoldGlobalUnitDimsBase<FoldGlobalUnitDimsPass> {
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    Explorer explorer(moduleOp, TraversalAction::RECURSE);
+    explorer.initialize();
+    IRRewriter rewriter(&getContext());
+    SymbolTable moduleSymbols(moduleOp);
+
+    // Rewrite globals without unit dims.
+    explorer.forEachGlobal([&](const Explorer::GlobalInfo *globalInfo) {
+      IREE::Util::GlobalOpInterface global = globalInfo->op;
+      auto tensorType = dyn_cast<RankedTensorType>(global.getGlobalType());
+      if (!tensorType || !global.isGlobalPrivate() ||
+          !global.isGlobalMutable()) {
+        return;
+      }
+      if (llvm::none_of(tensorType.getShape(),
+                        [](int64_t size) { return size == 1; })) {
+        return;
+      }
+      SmallVector<IREE::Util::GlobalLoadOpInterface> loadOps =
+          llvm::to_vector(globalInfo->getLoads());
+      SmallVector<IREE::Util::GlobalStoreOpInterface> storeOps =
+          llvm::to_vector(globalInfo->getStores());
+      if (failed(foldUnitDimsOnGlobal(rewriter, global, loadOps, storeOps,
+                                      moduleSymbols))) {
+        return signalPassFailure();
+      }
+    });
+
+    // Try to fold reshapes introduced by folding unit dims.
+    {
+      MLIRContext *context = &getContext();
```

Comment:
Can you declare it in the beginning of the function, so the `IRRewriter` can also use it?